### PR TITLE
Development testing clients

### PR DESCRIPTION
**Pull request by Dina-Abd-Elrahman**
Add `Docker` client tests.

### CODE REVIEW

File: tests/clients/docker_client.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        #. Create docker container.
+        #. Use stop method to stop docker container.
+        #. Check that docker is stopped correctly.
+        #. Use stop method to stop docker with non exist name, should raise an error.
+        """
+        self.info("Use stop method to stop docker container")
+        self.assertTrue(self.client.stop(self.DOCKER_NAME))
+
+        self.info("Check that docker is stopped correctly")
+        self.info("Check that docker is started correctly")
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:138]: Repeated line

File: tests/clients/docker_client.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            self.client.start(randint(1, 1000))
+            self.assertTrue("No such container" in error.exception.args[0])
+
+    def test04_docker_stop(self):
+        """
+        Test start docker.
+
+        **Test scenario**
+        #. Create docker container.
+        #. Use stop method to stop docker container.
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:130]: Need to check if it is running before stopping it.

File: tests/clients/docker_client.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        self.assertTrue(self.client.start(DOCKER_NAME))
+
+        self.info("Check that docker is started correctly")
+        output, error = self.os_command("docker inspect {} | grep \"Running\"".format(DOCKER_NAME))
+        self.assertIn("\"Running\": true", output.decode())
+
+        self.info("Use start method to start docker with non exist name, should raise an error")
+        with self.assertRaises(Exception) as error:
+            self.client.start(randint(1, 1000))
+            self.assertTrue("No such container" in error.exception.args[0])
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:122]: use `assertIn` instead

File: tests/clients/docker_client.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+        self.info("Check that docker is stopped correctly")
+        self.info("Check that docker is started correctly")
+        output, error = self.os_command("docker inspect {} | grep \"Running\"".format(self.DOCKER_NAME))
+        self.assertIn("\"Running\": false", output.decode())
+
+        self.info("Use start method to start docker with non exist name, should raise an error")
+        with self.assertRaises(Exception) as error:
+            self.client.stop(randint(1, 1000))
+            self.assertTrue("No such container" in error.exception.args[0])
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:145]: use `assertIn`

File: tests/clients/docker_client.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+        self.info("Create second container")
+        DOCKER_NAME_2 = "docker_{}".format(randint(100, 10000))
+        DOCKER_ID_2 = self.client.create(DOCKER_NAME_2, self.DOCKER_IMAGE).short_id
+
+        self.info("Use list subcommand with option all=True to list all containers")
+        self.info("Check the output of list command and make sure that it lists the two containers")
+        docker_list = self.client.list(all=True)
+        output, error = self.os_command("docker ps -qaf \"name=docker|DOCKER\"")
+        for i in range(len(docker_list)):
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:91]: Why not we loop on docker_list itself without len() and range()?

File: tests/clients/docker_client.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+            if docker_list[i].short_id in output.decode():
+                DOCKER_ID_1 = docker_list[i].short_id
+                self.assertIn(DOCKER_ID_1, output.decode())
+            elif docker_list[i].short_id == DOCKER_ID_2:
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:95]: what if it is not in the list?

File: tests/clients/docker_client.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        self.assertFalse(output.decode())
+
+        self.info("Create a running container")
+        DOCKER_NAME = "DOCKER_{}".format(randint(101, 1000))
+        self.client.run(DOCKER_NAME, self.DOCKER_IMAGE)
+
+        self.info("Try to delete the running docker using force=False option it should raise an error")
+        with self.assertRaises(Exception) as error:
+            self.client.delete(DOCKER_NAME, force=False)
+            self.assertTrue("Stop the container before attempting removal or force remove" in error.exception.args[0])
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:208]: use `assertIn`

File: tests/clients/docker_client.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        **Test scenario**
+        #. Create a docker container.
+        #. Use delete method to delete the container, with option force=True to delete the running container,
+           should be deleted correctly.
+        #. Check that the container has been deleted correctly.
+        #. Create a stopped container.
+        #. Try to delete the stopped docker using force=False should be deleted correctly.
+        #. Check that the container has been deleted correctly.
+        #. Create a running container.
+        #. Try to delete the running docker using force=False option it should raise an error.
```
Comment (AhmedHanafy725) [inline tests/clients/docker_client.py:179]: Need to check every time that the container is created before deleting it.

File: tests/clients/test_docker_client.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+        self.info("Try to get this container one time using container name")
+
+        docker = self.client.get(self.DOCKER_NAME)
+        self.assertIn(docker.short_id, output)
+
+        self.info("Try to get a container with a non valid name, and make sure that it raises an error")
+        with self.assertRaises(Exception) as error:
+            self.client.get("test_docker_{}".format(randint(100, 1000)))
+            self.assertTrue("No such container" in error.exception.args[0])
```
Comment (AhmedHanafy725) [inline tests/clients/test_docker_client.py:63]: use `assertIn`

File: tests/clients/test_docker_client.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        self.info("Check the output of list command")
+        self.info(" it should contain the first docker, but doesn't contain the second one")
+        self.assertIn(DOCKER_ID_1, str(docker_list))
+        self.assertNotIn(DOCKER_ID_2, str(docker_list))
+
+        self.info("Use list subcommand with option all=True to list all containers")
+        self.info("Check the output of list command and make sure that it lists the two containers")
+        docker_list = self.client.list(all=True)
+        self.assertIn(
+            DOCKER_ID_1 and DOCKER_ID_2,
```
Comment (AhmedHanafy725) [inline tests/clients/test_docker_client.py:102]: Adding two object with `and` result always in the second object, so this check only checking that DOCKER_ID_2 in the list.

File: tests/base_tests.py [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
 
 
 class BaseTests(TestCase):
     LOGGER = logger
 
     def setUp(self):
-        print('\t')
-        self.info('Test case : {}'.format(self._testMethodName))
+        print("\t")
+        self.info("Test case : {}".format(self._testMethodName))
```
Comment (abom) [inline tests/base_tests.py:12]: what's self.info? and what about j.logger? we can even create custom format for tests if needed.

**Review by abom — CHANGES_REQUESTED** [review]

**Pull request merged**